Code/Source/solver/mat_fun.h

```diff
@@ -61,20 +115,307 @@ namespace mat_fun {
 
     Array<double> mat_symm(const Array<double>& A, const int nd);
     Array<double> mat_symm_prod(const Vector<double>& u, const Vector<double>& v, const int nd);
+
+    template <int nsd>
+    Eigen::Matrix<double, nsd, nsd> mat_symm_prod_eigen(const Eigen::Matrix<double, nsd, 1>& u, const Eigen::Matrix<double, nsd, 1>& v) {
+        return 0.5 * (u * v.transpose() + v * u.transpose());
+    }
+
     double mat_trace(const Array<double>& A, const int nd);
 
     Tensor4<double> ten_asym_prod12(const Array<double>& A, const Array<double>& B, const int nd);
     Tensor4<double> ten_ddot(const Tensor4<double>& A, const Tensor4<double>& B, const int nd);
     Tensor4<double> ten_ddot_2412(const Tensor4<double>& A, const Tensor4<double>& B, const int nd);
     Tensor4<double> ten_ddot_3424(const Tensor4<double>& A, const Tensor4<double>& B, const int nd);
 
+    /**
+     * @brief Contracts two 4th order tensors A and B, C_ijkl = A_ijmn * B_klmn over the last two dimensions
+     * 
+     * @tparam nsd, the number of spatial dimensions
+     * @param A, the first 4th order tensor
+     * @param B, the second 4th order tensor
+     * @return Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>> 
+     */
+    template <int nsd>
+    Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>>
+    ten_ddot_eigen(const Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>>& A, 
```

Comment:
We don't need to prefix function names with `ten` and with an `eigen` suffix. 

It would be good to create a `Tensor` alias for `Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>>`